### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by MarcoCouto**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,13 @@
 # Project Vania - A Fair Distributor
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which (fairly) distributes a list of arbitrary **objects** through a set of **targets**.
 
-Using linear programming, this module takes into consideration the **weights** of the **targets** relative to the **objects** and distributes them in the **fairest way possible**.
+In a more pratical way, this module considers 3 key components:
+* **object**: some kind of entity that can be assigned to something.
+* **target**: the entity that will have one (or more) **objects** assigned to it.
+* **weight**: a map from **object** to **target**, which represents the cost of assigning the first to the second.
```
Comment (fntneves) [inline README.md]: *weights*. Is it really a map? It is actually a list of lists.
I think we could call it 'matrix'.

File: README.md [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
+
+Three lists are given as input to the module: **objects**, **targets** and **weights**.
+Using linear programming, the **weights** of the **targets** relative to the **objects** are taken into consideration and used to build the constraints of an Integer Linear Programming (ILP) model. An ILP solver is then used, in order to distributes the **objects** through the **targets**, in the *fairest way possible*.
```
Comment (fntneves) [inline README.md]: *in order to distribute*

File: README.md [hunk 3]
```diff
@@ -1,7 +1,13 @@
...
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which (fairly) distributes a list of arbitrary **objects** through a set of **targets**.
 
-Using linear programming, this module takes into consideration the **weights** of the **targets** relative to the **objects** and distributes them in the **fairest way possible**.
+In a more pratical way, this module considers 3 key components:
+* **object**: some kind of entity that can be assigned to something.
+* **target**: the entity that will have one (or more) **objects** assigned to it.
+* **weight**: a map from **object** to **target**, which represents the cost of assigning the first to the second.
+
+Three lists are given as input to the module: **objects**, **targets** and **weights**.
```
Comment (fntneves) [inline README.md]: The following three lists ...

File: README.md [hunk 4]
```diff
@@ -11,11 +17,11 @@ For instance, this module can be used to fairly distribute:
 ## Our Meaning of Fairness
 
 We define **Fairness** as:
- * The total weight for the resulting attribution of objects to targets should be minimal.
+ * The total cost (**weight**) of distributing all **objects** through the **targets** should be minimal.
 This enforces that the least amount of shared effort is made.
 
 _Optionally_, the following rule can be applied (enabled by default):
- * The difference between the total weight distributed between targets is minimal.
+ * The difference between the weight distributed by **targets** is minimal.
```
Comment (fntneves) [inline README.md]: the individual weight of each target is minimal.

**Review by fntneves — CHANGES_REQUESTED** [review]
See my comments.

Also, remove the following line from **Usage** section, or change it to describe the use case scenario.
*A quick example for 3 abstract targets, 4 abstract objects and the following weight matrix.*

File: README.md (hunk 1 above)
Comment (MarcoCouto) [reply README.md]: _weights_ is a matrix, but a **_weight_** is a map from an object to a target.
Here, I am only refering to the concept _weight_.

Comment (MarcoCouto) [reply README.md]: The input to the module is in a form of a matrix, but it could be something else in the future.
I think the _weight_ concept could be referred to as a map, and all _weights_ as a list of maps or something, don't you think? Or should the name **map** be switched to something else (like **association**)?

Comment (fntneves) [reply README.md]: Following your intention to define what weight is, I would say that it is *the cost ....*

Regarding your second comment, I think if we change the input type, then we'll update docs accordingly. For now, it is a list. What do you think?

Comment (MarcoCouto) [reply README.md]: Yes, _cost_ is a more understandable concept then _weight_. I will switch it.

Regarding the other issue, I think refering to _weights/costs_ as a list is not completely correct...
Lets try building a definition. First try:
- _weights/costs_ is a collection, which contains for each _target_ the _cost_ of assigning every _object_ to it, and is represented as a matrix.

What do you think?

Comment (fntneves) [reply README.md]: I like it! It is now very clear.
IMO, I would change to your suggestion!

Comment (MarcoCouto) [reply README.md]: Ok! Let me change the rest and I'll commit the changes

File: README.md [hunk 5]
```diff
@@ -4,24 +4,24 @@
 In a more pratical way, this module considers 3 key components:
 * **object**: some kind of entity that can be assigned to something.
 * **target**: the entity that will have one (or more) **objects** assigned to it.
-* **weight**: a map from **object** to **target**, which represents the cost of assigning the first to the second.
+* **cost**: a value associated between an **object** and a **target**, which represents the cost of assigning the first to the second.
```
Comment (fntneves) [inline README.md]: Change to this:

weight: represents the cost of assigning a given object to a target.

File: README.md [hunk 6]
```diff
@@ -40,26 +40,51 @@ To work with the source code, clone this repository:
...
+And here is the solution!
+```python
 # Output
 {
-    'user1': ['task1'], # User 1 does the task1
-    'user2': ['task2']  # User 2 does the task2
+    'Team A': ['Task 1'], 				# Team A does the Task 1
+    'Team C': ['Task 3', 'Task 4'], 	# Team B does the Tasks 3 and 4
+    'Team B': ['Task 2']}				# Team C does the Task 2
 }
```
Comment (fntneves) [inline README.md]: **Tasks**: Front-end Development, Back-end Development, Testing, Documentation.
Fix the alignment of comments.

What is your opinion about adding the full example in the end of this section?

Something like:

Here is the full example:
```python
...
```

**Review by fntneves — CHANGES_REQUESTED** [review]
Please revert costs to weights, except its definition.
Update tasks in Usage section.
Add full code to Usage section.

File: README.md (hunk 6 above)
Comment (MarcoCouto) [reply README.md]: I can agree on that. Maybe it turns out to be a little too repetitive, but we'll see.

Comment (fntneves) [reply README.md]: I understand, but this is good for "copy-paste" lovers :D

**Comment by fntneves** [comment]
For me, it's OK. I approve this. I've added @Hugainz as reviewer.

I suggest to wait for his review and then click at **squash and merge**. Afterwards, you can edit the description to remove unnecessary information, such as *(Almost)...* from commit message.

**Review by fntneves — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -1,21 +1,27 @@
...
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which (fairly) distributes a list of arbitrary **objects** through a set of **targets**.
 
-Using linear programming, this module takes into consideration the **weights** of the **targets** relative to the **objects** and distributes them in the **fairest way possible**.
+In a more pratical way, this module considers 3 key components:
+* **object**: some kind of entity that can be assigned to something.
+* **target**: the entity that will have one (or more) **objects** assigned to it.
+* **weight**: represents the cost of assigning a given **object** to a **target**.
+
+The following three lists are given as input to the module: **objects**, **targets** and **weights**.
```
Comment (hgg) [inline README.md]: I'd remove this sentece. See comments on line 4.

File: README.md [hunk 8]
```diff
@@ -1,21 +1,27 @@
 # Project Vania - A Fair Distributor
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which (fairly) distributes a list of arbitrary **objects** through a set of **targets**.
 
-Using linear programming, this module takes into consideration the **weights** of the **targets** relative to the **objects** and distributes them in the **fairest way possible**.
+In a more pratical way, this module considers 3 key components:
```
Comment (hgg) [inline README.md]: To simplify, I'd change this to 
> Pratically, this module considers the following components, which must be input:

File: README.md [hunk 9]
```diff
@@ -1,21 +1,27 @@
 # Project Vania - A Fair Distributor
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which (fairly) distributes a list of arbitrary **objects** through a set of **targets**.
```
Comment (hgg) [inline README.md]: I believe we want to clearly state that our distribution is fair. No need for braces :)

File: README.md [hunk 10]
```diff
@@ -1,21 +1,27 @@
...
+* **weight**: represents the cost of assigning a given **object** to a **target**.
+
+The following three lists are given as input to the module: **objects**, **targets** and **weights**.
+Using linear programming, the **weights** of the **targets** relative to the **objects** are taken into consideration and used to build the constraints of an Integer Linear Programming (ILP) model. An ILP solver is then used, in order to distribute the **objects** through the **targets**, in the *fairest way possible*.
 
 For instance, this module can be used to fairly distribute:
-* A set of tasks (objects) among a group of people (targets) according to their preferences to do each task (weights).
-* A set of projects (objects) among development teams (targets) according to their skill-level (weights) on the required skills for each project.
+* A set of tasks (objects) among a group of people (targets) according to their preferences to do each task (costs).
+* A set of projects (objects) among development teams (targets) according to their skill-level (costs) on the required skills for each project.
```
Comment (hgg) [inline README.md]: Where does the concept of costs come from? If **costs** is found to be a preferable nomenclature we should replace weights with costs altogether.

File: README.md (hunk 8 above)
Comment (fntneves) [reply README.md]: Does 'inputted' mean this ? 🤔 

At least it is not very common in README.
Maybe "must be given as input".

Comment (MarcoCouto) [reply README.md]: I think "inputted" is not correct. At least, I never saw it before.

File: README.md (hunk 9 above)
Comment (MarcoCouto) [reply README.md]: I agree. Yet, this is the first time the term appears, and we haven't define it yet. The ideia behind the braces is "ok, this is a fair distributor, but you don't know what fair os yet!" Obviously, I can remove the braces tho, if you disagree.

File: README.md (hunk 10 above)
Comment (MarcoCouto) [reply README.md]: Yes, I missed it. It should be weights

File: README.md (hunk 8 above)
Comment (hgg) [reply README.md]: My mistake :)
Input is the word we are looking for. 

EDIT: I'm editing the first comment :)

File: README.md [hunk 11]
```diff
@@ -34,29 +40,70 @@ To work with the source code, clone this repository:
...
+    [1, 2, 3, 2],		# hours for Team A to complete each task
+    [3, 1, 4, 2],		# hours for Team B to complete each task
+    [3, 4, 1, 1]		# hours for Team C to complete each task
+]
+
+distributor = FairDistributor(targets, objects, weights)
+print(distributor.distribute())
+```
+
 ## Contributions and Bugs
```
Comment (hgg) [inline README.md:107]: What do you think about this section (**Contribution and Bugs**)? I believe it could be improved. 
My suggestion is to separate the text into more paragraphs and maybe rephrase/restructure them.

File: README.md (hunk 9 above)
Comment (hgg) [reply README.md]: I get your point, but I'm not sure the braces add anything to the message. What do you think of making the word **fairly** a link to the section where the concept is  explained?

File: README.md (hunk 11 above)
Comment (fntneves) [reply README.md:107]: It needs to be improved, not only its structure but also the content.

I'll take care of it in the future, since we must explain how to test things.

Comment (MarcoCouto) [reply README.md:107]: Honestly, I think it is very clear :)
Can you be more specific in what you would change?

File: README.md (hunk 9 above)
Comment (MarcoCouto) [reply README.md]: That is a great idea :) I'll do it ASAP.

File: README.md (hunk 11 above)
Comment (hgg) [reply README.md:107]: Ok @fntneves.
@MarcoCouto I suggest adding the missing paragraph between line 63 and 64 and leaving it then.
I'll sum up and request all changes.

**Review by hgg — CHANGES_REQUESTED** [review]
As discussed, it would be great to have the following changes:
* Remove braces from the word **fairly** in line 2 and make it a link to  **Our Meaning of Fairness** section.
* Change line 4 to "Pratically, this module considers the following components, which must be input:"
* Remove line 9.
* Change costs to weights in lines 13 and 14.
* Add an empty line (paragraph) between line 110 and 111.

File: README.md [hunk 12]
```diff
@@ -34,29 +40,70 @@ To work with the source code, clone this repository:
...
-
+To start using the **Fair Distributor**, you need first to import it, by doing this:
 ```python
 from vania.fair_distributor import FairDistributor
+```
+Now, just feed it with your problem variables, and ask for the solution.
+To better explain how you can do it, lets consider a specific example.
+
+Suppose that you are managing a project, which contains **4** tasks: _Front-end Development_, _Back-end Development_, _Testing_, and _Documentation_.
+need to assign **4** tasks through a set of **3** teams: _A_, _B_ and _C_.
```
Comment (andrebp) [inline README.md]: You should change the "need" to "And you need to".

File: README.md [hunk 13]
```diff
@@ -34,29 +40,70 @@ To work with the source code, clone this repository:
...
+|--------|-----------------------|----------------------|---------|---------------|
+|_Team A_|          1h           |          2h          |    3h   |       2h      |
+|_Team B_|          3h           |          1h          |    4h   |       2h      |
+|_Team C_|          3h           |          4h          |    1h   |       1h      |
 
-targets = ['user1', 'user2']
-objects = ['task1', 'task2']
+Here, we consider tasks as **objects**, teams as **targets** and the hours expressed in each cell are the **weights**.
+
+It is necessary to create a data strcuture for each component. **Objects** and **targets** are lists, while **weights** is a collection, which contains for each target the cost of assigning every object to it, and is represented as a matrix.
```
Comment (andrebp) [inline README.md]: You should change "weights is a " to "weights are a ", it's plural :) .

Comment (fntneves) [reply README.md]: @MarcoCouto "strcuture"

@andrebrandao21 **weights** is the parameter, not the collection.

File: README.md (hunk 8 above)
Comment (MarcoCouto) [reply README.md]: I still don't like the _Pratically_ at the beginning of the sentence.
What do you think about this: 
```
To be more explicit, this module considers 3 key components:
* object* : ...
* target* : ...
* weight* : ...
A list of each of this components must be given as input to ...
```
?

File: README.md (hunk 12 above)
Comment (MarcoCouto) [reply README.md]: Something happened there... The sentence shouldn't start like that. I'll propose something a bit different ;)

File: README.md (hunk 8 above)
Comment (MarcoCouto) [reply README.md]: Well... Maybe `A collection of each...`, since `weights` is a matrix, thus is also a collection.

**Review by hgg — CHANGES_REQUESTED** [review]
We're almost there :+1: 
I'd just like to ask you to fix two typos:
* Line 9 : **this** should be **these**
* Line 12: **distribute** 

Also, as a suggestion only:
On Line 4, change **To be more explicit** to **Simply put**

File: README.md (hunk 8 above)
Comment (fntneves) [reply README.md]: What about to simply say *This module considers...* ?

**Review by hgg — CHANGES_REQUESTED** [review]
I'm really sorry I missed this last time. 
Line 9: _...these components **are** given as..._
I believe this will be the last change :)

**Comment by fntneves** [comment]
@Hugainz I am pretty sure that it is correct. It is related to the collection, not to its components. Please confirm.

**Comment by MarcoCouto** [comment]
Exactly :)

File: README.md [hunk 14]
```diff
@@ -1,7 +1,13 @@
 # Project Vania - A Fair Distributor
-**Fair Distributor** is module that allows fair distribution of any number of **objects** through a group of **targets**.
+**Fair Distributor** is a module which [fairly](our-meaning-of-fairness) distributes a list of arbitrary **objects** through a set of **targets**.
```
Comment (fntneves) [inline README.md]: This link is not working. Fix it by adding a # at the beginning of the target section.

`[fairly](#our-meaning-of-fairness)`

**Review by fntneves — CHANGES_REQUESTED** [review]
Fix link to the section about the meaning of fairness.

**Review by fntneves — APPROVED** [review]

**Review by hgg — APPROVED** [review]
You're right. I'm accepting this :+1:
